### DISCUSSION TITLE
Converting Stack of Tiff images to OBJ file for use in Unity

### DISCUSSION THREAD

**Discussion by ojrade** [pyvista/pyvista#4180] [Q&A]
Hi, 
I have a stack of tiff images representing an MRI scan, and I am trying to create and obj file in order to use in Unity. Would anybody be able to lead me down how to upload the files into python, and create the model or if its even possible? 

I know there are programs that can do this like 3D slicer which also utilizes VTK, but I want to be able to do it automated through python if possible.
Thanks for the help!

**Comment by banesullivan** [comment]
My workflow would be:

1. Load each tiff/Z-layer and stack into a single `vtkImageData` (`pyvista.UniformGrid`) -- could also stack into a single 3D NumPy array then pass to PyVista
2. Perform some sort of geometry extraction to get the surface you want to export to OBJ format. This could be a `threshold()` filter, etc.
3. Extract surface: `.extract_surface()`
4. Save: `pv.save_meshio('foo.obj', surf)`

The plotter also has an export OBJ utility if you want to do this differently than extracting an isosurface: https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.export_obj.html#export-obj

**Reply by MatthewFlamm** [reply]
> 1. Load each tiff/Z-layer and stack into a single `vtkImageData` (`pyvista.UniformGrid`) -- could also stack into a single 3D NumPy array then pass to PyVista

This can be done with [`pyvista.TIFFReader`](https://docs.pyvista.org/api/readers/_autosummary/pyvista.TIFFReader.html) or more simple with [`pyvista.read`](https://docs.pyvista.org/api/utilities/_autosummary/pyvista.read.html) to get single `UniformGrid` 'slices' if you have multiple files.  Then you can combine them into a single 3D `UniformGrid` if indeed it is a uniform grid, or a different type if nonuniform.

> 4. Save: pv.save_meshio('foo.obj', surf)

We should allow this directly as `mesh.save('foo.obj')` in pyvista.  It is available in vtk: https://vtk.org/doc/nightly/html/classvtkOBJWriter.html

**Reply by banesullivan** [reply]
Agreed. We currently use `vtkOBJExporter` with the `Plotter` but should definitely add `vtkOBJWriter` support

**Reply by ojrade** [reply]
I am trying to get threshold to work while using google colab but I just crash every time I run the plot method. Any suggestions why? 

import pyvista as pv
import os
import glob
import tifffile as tiff

dir_path = "/content/drive/MyDrive/3DsegVR/brain/"
tiff_files = sorted([os.path.join(dir_path, file) for file in os.listdir(dir_path)
                    if file.endswith(".tif")
                    and not file.endswith(".tif.meta")])
data = tiff.imread(tiff_files[0])
dims = data.shape
grid = pv.UniformGrid()
grid.dimensions = dims
for i, filename in enumerate(tiff_files):
    data = tiff.imread(filename)
    grid.point_data[f"Layer{i}"] = data.flatten(order="F")
threshed = grid.threshold(value=0.02)
threshed.plot(show_scalar_bar=False, show_edges=True)

This might be kind of obvious, but I am quite new to this kind of work and have spent a while working it. Any help on how exactly geometry extraction stuff works would also be great.

**Reply by banesullivan** [reply]
PyVista is not well supported on Colab, ref: https://github.com/pyvista/pyvista/issues/3851

You're likely getting segfaults/kernel crashes because the GL dependencies are not properly configured in your environment

**Reply by ojrade** [reply]
Hi again, I am trying to run this code now in a local environment and everything is running, but the model generated looks like it only has 2 layers when it should have 23. Is there something wrong with how my grid is being setup?

import pyvista as pv
import os
import glob
import tifffile as tiff

dir_path = "/content/drive/MyDrive/3DsegVR/brain/"
tiff_files = sorted([os.path.join(dir_path, file) for file in os.listdir(dir_path)
if file.endswith(".tif")
and not file.endswith(".tif.meta")])
data = tiff.imread(tiff_files[0])
dims = data.shape
grid = pv.UniformGrid()
grid.dimensions = dims

for i, filename in enumerate(tiff_files):
    data = tiff.imread(filename)
    grid.point_data[str(i)] = data.flatten(order="F")
grid.spacing = (1,1,10)
print(grid)

threshed = grid.threshold_percent((0.2,0.4))
surf = threshed.extract_surface()
grid.plot(cmap='gist_earth_r', show_scalar_bar=False, show_edges=True)
print(surf)
pv.save_meshio('test.obj', surf)

this is what my printed grid looks like
UniformGrid (0x1b73ed68d60)
  N Cells:	130050
  N Points:	196608
  X Bounds:	0.000e+00, 2.550e+02
  Y Bounds:	0.000e+00, 2.550e+02
  Z Bounds:	0.000e+00, 2.000e+01
  Dimensions:	256, 256, 3
  Spacing:	1.000e+00, 1.000e+00, 1.000e+01
  N Arrays:	23



image for reference:
![brokemodel](https://user-images.githubusercontent.com/48012294/230163099-050f2474-3977-4b90-bf3b-67e20526847e.PNG)
